x-pack/plugins/actions/server/saved_objects/index.ts

```diff
@@ -69,6 +71,17 @@ export function setupSavedObjects(
     hidden: true,
     namespaceType: 'single',
     mappings: mappings.action_task_params as SavedObjectsTypeMappingDefinition,
+    deleteOnUpgrade: async (client) => {
+      const oldestIdleActionTask = await getOldestIdleActionTask(client, taskManagerIndex);
```

Comment:
**question:** Does `getOldestIdleActionTask` query run on the old (previous) task manager index? or is there a chance the task manager index gets migrated first and impacts the result returned by this function?

I'm not sure how the internals of the migrations system works, so this may be a non-issue 🙂

Reply:
> Does getOldestIdleActionTask query run on the old (previous) task manager index?

The queries are run on the original source index (previous index from which we are migrating).

> or is there a chance the task manager index gets migrated first and impacts the result returned by this function?

Great question! Each index is migrated independently of the state of any others but I'm not sure if there is a set order in which they are migrated.

If we only have one Kibana instance and it runs the migrations and any others are shut down, then the risk of tasks still running against an index that's being migrated is lower than if other (pre-migrated) instances still accepting traffic. 

While we recommend all other instances be shut down while a single node handles the upgrade, I know that rolling upgrades do happen.

Reply:
Migrations for each index are run independently at this time and they are executed in parallel. Because of this, there is a risk that the `deleteOnUpgrade` hook for the `action_tasks_params` (the type that is on the `.kibana` index) may get executed _after_ the `.kibana_task_manager` index has already been migrated. There are two cases where this could happen:
- On a single Kibana instance deployment:
	- The `.kibana_task_manager` index 'wins the race' and finishes migrating before the `deleteOnUpgrade` hook is called for the `action_tasks_params` type. 
	- In this case, the task manager itself has not actually started executing because the migrations must be complete before Core will call the plugins' `start()` methods.
	- This should not pose a problem and should result in the same 
- On a multiple Kibana instance deployment:
	- One node completes the migration of both indices and calls the plugin `start()` methods. This node 'wins' the migration of both indices.
	- Another node that was launched at the same time hangs in the middle of the migration process (due to slow network, CPU scheduling, etc.) and calls the `deleteOnUpgrade` hook _after_ the other node's task manager has already started processing tasks. This could result in a different timestamp (more recent) & filter being returned from `deleteOnUpgrade`.
	- The second node will actually 'fail' on the next step when it attempts to index documents to the temporary index because the other node would have already set it's write block (or even deleted the temporary index). Thus this newer timestamp should never actually affect the final migration state. This 'failure' should not actually cause the second node to fail upgrading, the algo should assume another node completed the migration and just skip ahead to starting Kibana.

In summary, I believe this current design will not result in different results based on any race conditions of how the indices are queries and migrated. We never start the task manager until all indices are done migrating on at least one node. Any migration attempts that happen after the task manager has been started on any node should fail and be short-circuited before writing any documents to the final target index.

Reply:
>Yes, the queries are run on the original source index (previous index from which we are migrating).

As long as this ☝️ remains true, I don't have a concern 🙂 the remainder of the above all makes sense to me as well.

I was mostly worried that a `timestamp` would result to "now" if `.kibana_task_manager` wins the race causing the query in `getOldestIdleActionTask` for `action_tasks_params ` to run on the migrated task manager index.

Reply:
> > Yes, the queries are run on the original source index (previous index from which we are migrating).
> 
> As long as this ☝️ remains true, I don't have a concern 🙂 the remainder of the above all makes sense to me as well.
> 
> I was mostly worried that a `timestamp` would result to "now" if `.kibana_task_manager` wins the race causing the query in `getOldestIdleActionTask` for `action_tasks_params ` to run on the migrated task manager index.

In the current implementation it is possible that the query is run against the already migrated task manager index, however, in this case either:
- The task manager hasn't been started yet (in the single-node case); or
- The task manager has started on another node, which must mean the entire migration has been completed and the results of this query will be essentially ignored (because we'll short-circuit ahead).

I'll test this manually, and if I can easily, add an integration test.

Reply:
I answered my own question, the query in `getOldestIdleActionTask` looks at `idle` tasks and not `failed` tasks, so the query to get the timestamp would still apply to the same data in the new index and calculate the same timestamp regardless of which index got migrated first.